src/main/java/com/gtnewhorizons/retrofuturabootstrap/asm/SafeAsmClassWriter.java

```diff
@@ -31,4 +36,113 @@ protected ClassLoader getClassLoader() {
         final ExtensibleClassLoader launchLoader = Main.launchLoader;
         return launchLoader != null ? launchLoader.asURLClassLoader() : Main.compatLoader;
     }
+
+    @Override
+    protected String getCommonSuperClass(String type1, String type2) {
+        ClassLoader classLoader = getClassLoader();
+        if (!(classLoader instanceof ExtensibleClassLoader)) {
+            Main.logger.warn(
+                    "SafeAsmClassWriter#getClassLoader() didn't return an ExtensibleClassLoader, falling back to default getCommonSuperClass implementation");
+            return super.getCommonSuperClass(type1, type2);
+        }
+        ExtensibleClassLoader loader = (ExtensibleClassLoader) classLoader;
+
+        // Crash if types aren't real
+        InheritanceNode inheritance1 = getInheritanceInfo(loader, type1);
+        if (inheritance1 == null) {
+            throw new TypeNotPresentException(type1, null);
+        }
+
+        InheritanceNode inheritance2 = getInheritanceInfo(loader, type2);
+        if (inheritance2 == null) {
+            throw new TypeNotPresentException(type2, null);
+        }
+
+        // Assignable checks, replicates the ones in asm
+        if (inheritance1.isAssignableFrom(inheritance2)) {
+            return type1;
+        }
+        if (inheritance2.isAssignableFrom(inheritance1)) {
+            return type2;
+        }
+
+        // this is a bit weird but it's what ASM does
+        if (inheritance1.accessor.isInterface() || inheritance2.accessor.isInterface()) {
+            return "java/lang/Object";
+        } else {
+            do {
+                inheritance1 = inheritance1.superClass;
+                // We might have some cut-off inheritance trees that don't properly reach back to Object. Avoid crashing
+                // on those.
+                if (inheritance1 == null) {
+                    return "java/lang/Object";
+                }
+            } while (!inheritance1.isAssignableFrom(inheritance2));
+            return inheritance1.accessor.binaryThisName();
+        }
+    }
+
+    private @Nullable InheritanceNode getInheritanceInfo(ExtensibleClassLoader loader, String type) {
+        InheritanceNode node = inheritanceCache.get(type);
+        if (node == null) {
```

Comment:
If you flip this condition to `if (node != null) {return node;}`, it avoids the long indent